Breeding-Insight/brapi-Java-TestServer
# BI-1945 Store Additional Info as JSONB (Hackathon)

### Description

[Jira story](https://breedinginsight.atlassian.net/browse/BI-1945).

- Pulled in Peter Selby's latest work which adds flyway.
    - The `spring.jpa.hibernate.ddl-auto` property in application.properties should be set to `validate`, as flyway is now responsible for generating and managing the schema.
- Updated entity classes and database schema to store `additionalInfo` as a JSONB column on each primary entity.
    - The `JsonbConverter` class is needed because Hibernate doesn't support JSONB columns out-of-the box.
- Added `spring.datasource.hikari.data-source-properties.stringtype=unspecified` to `application.properties.template`, make sure this is copied to your local `application.properties`!
    - Without this, the following error will be thrown: `org.postgresql.util.PSQLException: ERROR: column "additional_info" is of type jsonb but expression is of type character varying`

> Note: after surveying groups currently using the BrAPI Java TestServer, it was determined that it would not be worth the substantial effort to try to support existing data with a migration.

[PR for configuration changes](https://github.com/Breeding-Insight/bi-docker-stack/pull/34) for deployments with bi-docker-stack.

### Running locally
Your local `application.properties` should look something like this:
```application.properties
server.port = 6081
server.servlet.context-path=/brapi/v2

spring.datasource.url=jdbc:postgresql://localhost:5432/demo
spring.datasource.hikari.data-source-properties.stringtype=unspecified
spring.datasource.username=postgres
spring.datasource.password=postgres

spring.datasource.driver-class-name=org.postgresql.Driver

spring.flyway.locations=classpath:db/migration,classpath:org/brapi/test/BrAPITestServer/db/migration
spring.flyway.schemas=public
spring.flyway.baselineOnMigrate=true

spring.jpa.hibernate.ddl-auto=validate
spring.jpa.show-sql=false

spring.mvc.dispatch-options-request=true

security.oidc_discovery_url=https://example.com/auth/.well-known/openid-configuration
security.enabled=false
```
Notes:
- the `spring.datasource.hikari.data-source-properties.stringtype=unspecified`, was needed to get JPA to handle jsonb.
- because the schema changes won't allow this branch to run against the database used by develop, a different database is targeted ("demo").
- port 6081 is used to prevent interference with a local dockerized brapi-server running on port 6080.

### Testing

1. Get bi-api and bi-web running as usual locally (brapi-server docker container can be running, this branch will use a different port).
2. Create an empty database (`docker exec -it bidb psql -U postgres -c 'CREATE DATABASE demo;'`
3. Checkout this branch (`git checkout feature/BI-1945`).
4. Modify your local application.properties (see example above).
5. Run maven `clean install`.
7. Run brapi-server. Flyway will create the schema in the "demo" database and run migrations.
8. Insert missing species into local brapi-server using the SQL below.
10. Login to DeltaBreed and create a new program using `http://localhost:6081` as the custom data storage location and test every DeltaBreed workflow you can think of.

Insert missing species:
```sql
INSERT INTO crop (auth_user_id, id, crop_name) VALUES ('anonymousUser', '4', 'Blueberry') ON CONFLICT DO NOTHING;
INSERT INTO crop (auth_user_id, id, crop_name) VALUES ('anonymousUser', '5', 'Salmon') ON CONFLICT DO NOTHING;
INSERT INTO crop (auth_user_id, id, crop_name) VALUES ('anonymousUser', '6', 'Grape') ON CONFLICT DO NOTHING;
INSERT INTO crop (auth_user_id, id, crop_name) VALUES ('anonymousUser', '7', 'Alfalfa') ON CONFLICT DO NOTHING;
INSERT INTO crop (auth_user_id, id, crop_name) VALUES ('anonymousUser', '8', 'Sweet Potato') ON CONFLICT DO NOTHING;
INSERT INTO crop (auth_user_id, id, crop_name) VALUES ('anonymousUser', '9', 'Trout') ON CONFLICT DO NOTHING;
INSERT INTO crop (auth_user_id, id, crop_name) VALUES ('anonymousUser', '10', 'Soybean') ON CONFLICT DO NOTHING;
INSERT INTO crop (auth_user_id, id, crop_name) VALUES ('anonymousUser', '11', 'Cranberry') ON CONFLICT DO NOTHING;
INSERT INTO crop (auth_user_id, id, crop_name) VALUES ('anonymousUser', '12', 'Cucumber') ON CONFLICT DO NOTHING;
INSERT INTO crop (auth_user_id, id, crop_name) VALUES ('anonymousUser', '13', 'Oat') ON CONFLICT DO NOTHING;
INSERT INTO crop (auth_user_id, id, crop_name) VALUES ('anonymousUser', '14', 'Citrus') ON CONFLICT DO NOTHING;
INSERT INTO crop (auth_user_id, id, crop_name) VALUES ('anonymousUser', '15', 'Sugar Cane') ON CONFLICT DO NOTHING;
INSERT INTO crop (auth_user_id, id, crop_name) VALUES ('anonymousUser', '16', 'Strawberry') ON CONFLICT DO NOTHING;
INSERT INTO crop (auth_user_id, id, crop_name) VALUES ('anonymousUser', '17', 'Honey') ON CONFLICT DO NOTHING;
INSERT INTO crop (auth_user_id, id, crop_name) VALUES ('anonymousUser', '18', 'Pecan') ON CONFLICT DO NOTHING;
INSERT INTO crop (auth_user_id, id, crop_name) VALUES ('anonymousUser', '19', 'Lettuce') ON CONFLICT DO NOTHING;
INSERT INTO crop (auth_user_id, id, crop_name) VALUES ('anonymousUser', '20', 'Cotton') ON CONFLICT DO NOTHING;
```
